go-apps/meep-tc-engine/ip-mgr.go

```diff
@@ -0,0 +1,244 @@
+	im.podIpMap = make(map[string]string)
+	im.svcIpMap = make(map[string]string)
+
+	log.Info("Successfully create IP Manager")
+	return im, nil
+}
+
+// SetPodList - Set list of pods to monitor IP addresses for
+func (im *IpManager) SetPodList(podList map[string]bool) {
+	im.mutex.Lock()
+	defer im.mutex.Unlock()
+
+	// Remove stale entries
+	for podName := range im.podIpMap {
+		if _, found := podList[podName]; !found {
+			delete(im.podIpMap, podName)
+		}
+	}
+
+	// Add missing entries
+	for podName := range podList {
+		if _, found := im.podIpMap[podName]; !found {
+			im.podIpMap[podName] = IP_ADDR_NONE
+		}
+	}
+}
+
+// SetSvcList - Set list of services to monitor IP addresses for
+func (im *IpManager) SetSvcList(svcList map[string]bool) {
+	im.mutex.Lock()
+	defer im.mutex.Unlock()
+
+	// Remove stale entries
+	for svcName := range im.svcIpMap {
+		if _, found := svcList[svcName]; !found {
+			delete(im.svcIpMap, svcName)
+		}
+	}
+
+	// Add missing entries
+	for svcName := range svcList {
+		if _, found := im.svcIpMap[svcName]; !found {
+			im.svcIpMap[svcName] = IP_ADDR_NONE
+		}
+	}
+}
+
+// GetPodIp - Get Pod IP address
+func (im *IpManager) GetPodIp(podName string) string {
+	podIp, found := im.podIpMap[podName]
+	if !found {
+		podIp = IP_ADDR_NONE
+	}
+	return podIp
+}
+
+// GetPodIp - Get Svc IP address
+func (im *IpManager) GetSvcIp(svcName string) string {
+	svcIp, found := im.svcIpMap[svcName]
+	if !found {
+		svcIp = IP_ADDR_NONE
+	}
+	return svcIp
+}
+
+// Start - Start monitoring IP addresses
+func (im *IpManager) Start() error {
+	im.mutex.Lock()
+	defer im.mutex.Unlock()
+
+	log.Debug("Starting IP Address Manager: ", im.name)
+
+	// Make sure ticker is not already running
+	if im.ticker != nil {
+		return errors.New("Ticker already running")
+	}
+
+	// Start ticker to periodically retrieve platform information
+	im.ticker = time.NewTicker(DEFAULT_TICKER_INTERVAL_MS * time.Millisecond)
+	go func() {
+		for range im.ticker.C {
+			im.mutex.Lock()
+
+			// Refresh IP addresses
+			im.refreshIpAddresses()
+
+			im.mutex.Unlock()
+		}
+	}()
+
+	return nil
+}
+
+// Stop - Stop monitoring IP addresses
+func (im *IpManager) Stop() {
+	im.mutex.Lock()
+	defer im.mutex.Unlock()
+
+	if im.ticker != nil {
+		im.ticker.Stop()
+		im.ticker = nil
+	}
+}
+
+// Refresh - Request a IP addresses
+func (im *IpManager) Refresh() {
+	im.mutex.Lock()
+	defer im.mutex.Unlock()
+
+	im.refreshIpAddresses()
+}
+
+// refreshIpAddresses - Refresh Pod & Service IP addresses
+func (im *IpManager) refreshIpAddresses() {
+	var err error
+	updated := false
+
+	// Establish connection to K8s API Server if not connected
+	if !im.isConnected {
+		im.clientset, err = connectToAPISvr()
+		if err != nil {
+			log.Error("Failed to connect with k8s API Server. Error: ", err)
+			return
+		}
+		im.isConnected = true
+	}
+
+	// Retrieve all sandbox pods from k8s api
+	podList, err := im.clientset.CoreV1().Pods(im.sandboxName).List(metav1.ListOptions{})
```

Comment:
I believe the above query will return all pods in the sandbox (core & scenario)
e.g. `kubectl get pods -n <sandboxName>`

For efficiency, it should only focus at the scenario pods
e.g ` kubectl get pods -n <sandboxName> -l meepOrigin=scenario`

Reply:
Agreed. Modifying as suggested. Note that the meepOrigin label is inserted in every deployment & statefulset (and hence the pods) using the meep-webhook, which guarantees that the label will be present.